### PR TITLE
listing/selecting disclosures using explicit dot-notation jsonpath

### PR DESCRIPTION
**Pull request by ragnika**
Usage example in `docs/sd-jsonpath.md`

This only adds support for explicit dot-notation jsonpath.

If the path provided doesn't start with `$.`, it is assumed to be from the root.

@vijayshiyani can you have a quick look and see if there's any glaring problem where we won't be able to use this with the sd-jwt-vc library

### CODE REVIEW

**Review by vijayshiyani — APPROVED** [review]

File: src/jsonpath.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
+import { splitSDJWT } from './common';
+import { SD_DIGEST, SD_LIST_PREFIX } from './constants';
+import { combineSDJWT, decodeDisclosure, decodeJWT, isObject } from './helpers';
+import { Disclosure, GetHasher, Hasher } from './types';
+
+export type DigestMap = Map<string, Disclosure>;
+
+type getDisclosuresOptions = {
+  ignoreInvalid?: boolean;
```
Comment (linasi) [inline src/jsonpath.ts:9]: Ignore invalid what? Could you rename it to something more descriptive?

File: src/jsonpath.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+};
+
+/**
+ * List and Retrieve Disclosures using
+ * Explicit dot-notation JSONpath expression
+ */
+export class SdJwtDisclosureParser {
```
Comment (linasi) [inline src/jsonpath.ts:16]: If I understand everything correctly, it is a tree data structure. Maybe the name could reflect it somehow? `SdJwtDisclosureTree`, `SdJwtDisclosureMap`, `SdJwtDisclosureStruct`?
I'm just thinking out loud here.

File: src/jsonpath.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+    if (parent) {
+      collection.push(parent);
+    }
+
+    return this.getParentDigests(parent, collection);
+  }
+}
+
+export class SDJsonpath {
```
Comment (linasi) [inline src/jsonpath.ts:153]: Would be nice to have some consistency in naming of the classes/functions. Here `SD` is both capitals. In other cases it is written as `Sd`, `SdJwt`, `SDJWT`. Could you please pick one and update the codebase?

File: src/test-utils/jsonpath.params.ts
Comment (linasi) [inline src/test-utils/jsonpath.params.ts:1]: Please rename file to something that would indicate mock data being inside. `jsonpath.mock-data.ts` or `jsonpath.test-data.ts` for example?

File: src/jsonpath.spec.ts [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+});
+
+const hasher = (data) => {
+  const digest = crypto.createHash('sha256').update(data).digest();
+  return base64encode(digest);
+};
+
+const getHasher: GetHasher = () => Promise.resolve(hasher);
+
+describe('#SDJsonpath', () => {
```
Comment (linasi) [inline src/jsonpath.spec.ts:619]: I would consider moving `SDJsonpath` implementation and tests into separate files.

**Review by linasi — APPROVED** [review]
Looks good. Take a look at those few comments I have left before merging it please.

File: src/helpers.ts [hunk 5]
```diff
@@ -56,31 +56,33 @@ export function decodeJWT(input: string): UnverifiedJWT {
...
 
-    return {
-      disclosure: d,
-      key,
-      value,
-    };
-  });
+  // if disclosure is a value in an array
+  // [<SALT>, <VALUE>]
+  if (decoded.length == 2) {
```
[the remark this answers is not in the document]
Comment (ragnika) [reply src/helpers.ts:71]: good idea

File: src/jsonpath.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+  // Map<childDigest, parentDigest>
+  private parentMap: Map<string, string> = new Map();
+
+  constructor({ jwt, digestMap }: { jwt: object; digestMap: DigestMap }) {
+    this.jwt = jwt;
+    this.digestMap = digestMap;
+  }
+
+  public parse() {
```
[the remark this answers is not in the document]
Comment (ragnika) [reply src/jsonpath.ts:33]: yeah, this is redundant. At one point, the parse was an async function so it was a separate step.

**Pull request merged**